src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -226,25 +222,26 @@ private void RemoveLRUs()
+            // Setting _eventQueueTaskState to EventQueueTaskDoNotStop here will force the event queue task in EventQueueTaskAction to continue even if the event queue is empty and it is ready to exit.
+            // It is mainly to prevent a rare (but theoretically possible) thread synchronization issue.
             // For example:
-            //   1. the task execution in EventQueueTaskAction() checks _eventQueueTaskStopTime and it has already passed (ready to exit)
-            //   2. the execution is switched to this thread (before it calls the Interlocked.CompareExchange() to set the _eventQueueTaskState to EventQueueTaskStopped)
-            //   3. the _eventQueueTaskStopTime is extended but it can't stop the task from stopping as the task has already passed the time check
-            //   4. now since the _eventQueueTaskState == EventQueueTaskRunning, it can be set to EventQueueTaskDoNotStop by the Interlocked.CompareExchange() below
-            //   5. if _eventQueueTaskState is successfully set to EventQueueTaskDoNotStop, the Interlocked.CompareExchange() in the EventQueueTaskAction() will fail
-            //      and the task will continue the while loop and the new _eventQueueTaskStopTime will keep the task running
-            //   6. if _eventQueueTaskState is NOT set to EventQueueTaskDoNotStop because of context switch back to the EventQueueTaskAction() and the _eventQueueTaskState is
+            //   1. the task execution in EventQueueTaskAction() checks event queue and it is empty (ready to exit)
+            //   2. the execution is switched to this thread (before the event queue task calls the Interlocked.CompareExchange() to set the _eventQueueTaskState to EventQueueTaskStopped)
+            //   3. now since the _eventQueueTaskState == EventQueueTaskRunning, it can be set to EventQueueTaskDoNotStop by the Interlocked.CompareExchange() below
+            //   4. if _eventQueueTaskState is successfully set to EventQueueTaskDoNotStop, the Interlocked.CompareExchange() in the EventQueueTaskAction() will fail
+            //      and the task will continue the while loop and the new event will keep the task running
+            //   5. if _eventQueueTaskState is NOT set to EventQueueTaskDoNotStop because of context switch back to the EventQueueTaskAction() and the _eventQueueTaskState is
             //      set to EventQueueTaskStopped (task exits), then the second Interlocked.CompareExchange() below should set the _eventQueueTaskState to EventQueueTaskRunning
             //      and start a task again (though this scenario is unlikely to happen)
             //
-            // Though this kind of context switching is probably unlikely to happen, this hopefully will take care of it.
+            // Without the EventQueueTaskDoNotStop state, this execution of this thread could pass the second CompareExchange() call below after step 2 above. The event queue task would still exit
```

Comment:
I'm finding the wording on this comment a little confusing; is this trying to say that if we didn't have EventQueueTaskDoNotStop we'd run into a scenario where the task might not be started up properly when a new event is added to the queue?

Reply:
@mafurman, yes you are right, it could happen if we don't have EventQueueTaskDoNotStop, though the probability is tiny, only if the event queue task ALREADY checked the queue and it was empty and the new event was added AFTER that check.
Do you have suggestion on the comments to make it easier to understand?

Reply:
Got it. Here's how I would reword it:
"Without the EventQueueTaskDoNotStop state check below, steps (3) and (4) above will not execute. Step (5) will also not be hit, as the event queue is still running (even though the EventQueueTaskAction() method has already checked that the queue is empty and is about to stop the task). This method (StartEventQueueTaskIfNotRunning()) will return, the context will switch over to EventQueueTaskAction(), and the task will terminate. This means no new task would be started to process the newly added event.
This scenario is unlikely to happen, as it can only occur if the event queue task ALREADY checked the queue and it was empty and the new event was added AFTER that check."

Reply:
Feel free to use some/all of this if it makes more sense to you to;  this is just my opinion. :)

Reply:
Thank you for the comments, they definitely help explain the scenario more. I made some minor changes.